File: examples/FSharpDev/HelloWorld/CustomStepExecControlExample.fs
module FSharpDev.HelloWorld.CustomStepExecControlExample

open System.Threading.Tasks

open FSharp.Control.Tasks.NonAffine

open NBomber
open NBomber.Contracts
open NBomber.FSharp

let run () =

    let step1 = Step.create("step_1", fun context -> task {
        context.Logger.Information($"{context.StepName} invoked")
        do! Task.Delay(milliseconds 500)
        return Response.ok()
    })

    let step2 = Step.create("step_2", fun context -> task {
        context.Logger.Information($"{context.StepName} invoked")
        do! Task.Delay(milliseconds 500)
        return Response.ok()
    })

    let step3 = Step.create("step_3", fun context -> task {
        context.Logger.Information($"{context.StepName} invoked")
        do! Task.Delay(milliseconds 500)
        return Response.ok()
    })

    Scenario.create "scenario" [step1; step2; step3]
    |> Scenario.withoutWarmUp
    |> Scenario.withLoadSimulations [KeepConstant(copies = 1, during = seconds 10)]
    |> Scenario.withCustomStepExecControl(fun execControl ->
        // step_1 will never be invoked
        match execControl with
        | ValueSome exec when not exec.PrevStepResponse.IsError -> ValueSome "step_2"
        | _ -> ValueSome "step_3"
    )
    |> NBomberRunner.registerScenario
    |> NBomberRunner.run
    |> ignore

File: examples/CSharpDev/Features/CustomStepExecControlExample.cs
namespace CSharpDev.Features;

using System;
using System.Threading.Tasks;
using NBomber.Contracts;
using NBomber.CSharp;

public class CustomStepExecControlExample
{
    public static void Run()
    {
        var step1 = Step.Create("step_1", async context =>
        {
            context.Logger.Information($"{context.StepName} invoked");
            await Task.Delay(TimeSpan.FromSeconds(0.1));
            return Response.Ok();
        });

        var step2 = Step.Create("step_2", async context =>
        {
            context.Logger.Information($"{context.StepName} invoked");
            await Task.Delay(TimeSpan.FromSeconds(0.1));
            return Response.Ok();
        });

        var step3 = Step.Create("step_3", async context =>
        {
            context.Logger.Information($"{context.StepName} invoked");
            await Task.Delay(TimeSpan.FromSeconds(0.1));
            return Response.Ok();
        });

        var scenario = ScenarioBuilder
            .CreateScenario("my_scenario", step1, step2, step3)
            .WithoutWarmUp()
            .WithLoadSimulations(Simulation.KeepConstant(1, TimeSpan.FromSeconds(10)))
            .WithStepInterception(context =>
            {
                // step_1 will never be invoked
                if (context.IsSome)
                {
                    if (!context.Value.PrevStepResponse.IsError)
                        return ValueOption.Some("step_2");
                }
                return ValueOption.Some("step_3");
            });

        NBomberRunner
            .RegisterScenarios(scenario)
            .Run();
    }
}
